### ISSUE TITLE
Different metrics and loss values in model.fit and model.evaluate for the same batch of data

### ISSUE THREAD

**Issue by julj** [tensorflow/tensorflow#29964]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 1.14.0
- Python version: 3.6.8
- CUDA/cuDNN version: cuda 10.0, cudnn 7.5.1
- GPU model and memory: NVIDIA GeForce RTX 2070

**Describe the current behavior**

When performing the evaluation on a model I get values for metrics and loss that differ from those obtained at the last training epoch. 

The code and the output is given [below](#issuecomment-504010990).

**Describe the expected behavior**

Because the batch of data is the same and the metrics computed are the same, I expect to get exactly the same values.

**Comment by achandraa** [comment]
Will it be possible for you to provide a full code snippet that can help us to reproduce the issue. Thanks!

**Comment by julj** [comment]
Sure. Here is a simple code to reproduce the issue.
Note that I used only one big batch because I read somewhere that the metrics monitored during the training were computed after each batch update,

Code:
```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import mnist


config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)


#######################
# Get subset of MNIST #
#######################

batch_size = 1024
num_classes = 10
h, w = 28, 28

(x_train, y_train), (_, _) = mnist.load_data()
x_train = x_train[:1024]
x_train = x_train.reshape(x_train.shape[0], h, w, 1) # default tf format is NHWC
y_train = y_train[:1024]
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
    

###############
# build model #
###############

inputs = keras.Input(shape=(h, w, 1), batch_size=batch_size)
x = layers.Conv2D(32, (3, 3))(inputs)
x = layers.Activation('relu')(x)
x = layers.MaxPooling2D(pool_size=(3, 3), strides=(2,2))(x)
x = layers.Flatten()(x)
x = layers.Dense(100, activation='elu', name='logits')(x)
outputs = layers.Dense(num_classes, activation='softmax')(x)
    
model = tf.keras.Model(inputs=inputs, outputs=outputs)

model.summary()


################################################
# define training parameters and compile model #
################################################

model.compile(
    loss=keras.losses.categorical_crossentropy,
    optimizer=keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999),
    metrics=['accuracy']
)


#########
# train #
#########

print("***** Training *****")

model.fit(
    x_train,
    y_train,
    epochs=10,
    verbose=2
)


############
# Evaluate #
############

print("***** Evaluation *****")

ev = model.evaluate(x_train, y_train)
print(ev)
```
Output:
```
Epoch 1/10
2019-06-20 14:33:27.217399: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2019-06-20 14:33:27.394428: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
1024/1024 - 1s - loss: 40.9096 - acc: 0.1318
Epoch 2/10
1024/1024 - 0s - loss: 58.4202 - acc: 0.2168
Epoch 3/10
1024/1024 - 0s - loss: 49.5150 - acc: 0.3896
Epoch 4/10
1024/1024 - 0s - loss: 51.4337 - acc: 0.3301
Epoch 5/10
1024/1024 - 0s - loss: 39.2524 - acc: 0.4150
Epoch 6/10
1024/1024 - 0s - loss: 27.9284 - acc: 0.4648
Epoch 7/10
1024/1024 - 0s - loss: 18.2593 - acc: 0.5752
Epoch 8/10
1024/1024 - 0s - loss: 12.8214 - acc: 0.5957
Epoch 9/10
1024/1024 - 0s - loss: 8.0606 - acc: 0.6504
Epoch 10/10
1024/1024 - 0s - loss: 4.9667 - acc: 0.7031
***** Evaluation *****
1024/1024 [==============================] - 0s 25us/sample - loss: 5.3507 - acc: 0.6465
[5.350688457489014, 0.6464844]
```

**Comment by achandraa** [comment]
@julj : Thanks for the information. I have tried multiple times on Colab with TF version 1.14.0-rc1 but was able to get slightly better accuracy and slightly less loss during evaluation than during training. Please have a look on [Colab Link](https://colab.sandbox.google.com/drive/19YAG3jq9ZMxgAex91lZrozKG8uQd0FqG#scrollTo=fily-ATErvQW).

**Comment by julj** [comment]
The issue is not about the values themselves but about the difference between the values at the last training step and the values at the evaluation step. I expect them to be equal.

**Comment by eduardofv** [comment]
@julj I've also reproduced on Colab and found small differences. Furthermore, if trained a bit longer the differences get smaller and the Acc graph becomes more stable. https://colab.research.google.com/drive/1lGgPndW8pXFLaL4-9Ut66EnEUBXMfsUN

```
***** Training *****
Epoch 1/10
1024/1024 - 0s - loss: 56.3011 - acc: 0.0889
Epoch 2/10
1024/1024 - 0s - loss: 70.6232 - acc: 0.2344
Epoch 3/10
1024/1024 - 0s - loss: 73.6560 - acc: 0.2500
Epoch 4/10
1024/1024 - 0s - loss: 56.5223 - acc: 0.3340
Epoch 5/10
1024/1024 - 0s - loss: 45.7556 - acc: 0.3379
Epoch 6/10
1024/1024 - 0s - loss: 34.4217 - acc: 0.4746
Epoch 7/10
1024/1024 - 0s - loss: 30.5761 - acc: 0.5049
Epoch 8/10
1024/1024 - 0s - loss: 21.0374 - acc: 0.5127
Epoch 9/10
1024/1024 - 0s - loss: 13.5673 - acc: 0.6230
Epoch 10/10
1024/1024 - 0s - loss: 10.9355 - acc: 0.6162
***** Evaluation *****
1024/1024 [==============================] - 0s 30us/sample - loss: 9.8181 - acc: 0.6064
[9.818145751953125, 0.6064453]
```

For 30 epochs:
```
***** Training *****
Epoch 1/30
1024/1024 - 2s - loss: 49.0334 - acc: 0.1006
Epoch 2/30
1024/1024 - 0s - loss: 103.3966 - acc: 0.0938
Epoch 3/30
1024/1024 - 0s - loss: 72.9386 - acc: 0.2168
Epoch 4/30
1024/1024 - 0s - loss: 48.9603 - acc: 0.3711
Epoch 5/30
1024/1024 - 0s - loss: 40.9398 - acc: 0.4121
Epoch 6/30
1024/1024 - 0s - loss: 45.3177 - acc: 0.4219
Epoch 7/30
1024/1024 - 0s - loss: 39.7212 - acc: 0.4502
Epoch 8/30
1024/1024 - 0s - loss: 28.5958 - acc: 0.5381
Epoch 9/30
1024/1024 - 0s - loss: 22.0282 - acc: 0.6025
Epoch 10/30
1024/1024 - 0s - loss: 18.7421 - acc: 0.6221
Epoch 11/30
1024/1024 - 0s - loss: 16.2718 - acc: 0.6201
Epoch 12/30
1024/1024 - 0s - loss: 15.4877 - acc: 0.6133
Epoch 13/30
1024/1024 - 0s - loss: 13.6785 - acc: 0.6211
Epoch 14/30
1024/1024 - 0s - loss: 10.8438 - acc: 0.6445
Epoch 15/30
1024/1024 - 0s - loss: 8.0689 - acc: 0.6689
Epoch 16/30
1024/1024 - 0s - loss: 5.7298 - acc: 0.7168
Epoch 17/30
1024/1024 - 0s - loss: 4.0341 - acc: 0.7656
Epoch 18/30
1024/1024 - 0s - loss: 2.8764 - acc: 0.8135
Epoch 19/30
1024/1024 - 0s - loss: 2.3743 - acc: 0.8311
Epoch 20/30
1024/1024 - 0s - loss: 2.2624 - acc: 0.8516
Epoch 21/30
1024/1024 - 0s - loss: 2.3173 - acc: 0.8477
Epoch 22/30
1024/1024 - 0s - loss: 2.3842 - acc: 0.8369
Epoch 23/30
1024/1024 - 0s - loss: 2.3811 - acc: 0.8311
Epoch 24/30
1024/1024 - 0s - loss: 2.3094 - acc: 0.8330
Epoch 25/30
1024/1024 - 0s - loss: 2.2365 - acc: 0.8447
Epoch 26/30
1024/1024 - 0s - loss: 2.1789 - acc: 0.8496
Epoch 27/30
1024/1024 - 0s - loss: 2.1076 - acc: 0.8516
Epoch 28/30
1024/1024 - 0s - loss: 2.0055 - acc: 0.8535
Epoch 29/30
1024/1024 - 0s - loss: 1.8593 - acc: 0.8604
Epoch 30/30
1024/1024 - 0s - loss: 1.6750 - acc: 0.8682
***** Evaluation *****
1024/1024 [==============================] - 0s 26us/sample - loss: 1.4750 - acc: 0.8828
[1.474952220916748, 0.8828125]
```

**Comment by julj** [comment]
Because the batch of data is the same and the metrics computed are the same, I expect the values to be **exactly** the same. I will edit my first comment accordingly.

**Comment by eduardofv** [comment]
Sorry, I should have explained myself better. The main difference between training (fit) and test (evaluate) is that the former computes the feed forward step on the NN and then updates the weights via back-propagation while the latter only does the first part. So, as I understand, the training step modifies the model after computing the loss (and the associated metrics) so when you do evaluation the model is different than the one used to calculate training loss. As training advances and the model becomes more stable the difference tends to be smaller and that's what I tried to show on the example. Of course, two calls to evaluation with the same model and data yields the exact same results. Hope this helps!

**Comment by julj** [comment]
I am familiar with the forward/backward steps. It seems more logical to me (but maybe I am wrong) when training a model to show the metrics after the backward steps.
So in your last example I would interpret the loss and metrics values in

```
Epoch 30/30
1024/1024 - 0s - loss: 1.6750 - acc: 0.8682
```

as the values obtained after the last training step, i.e. the value for the model obtained on the training data.

In that case I would not understand why the evaluation on the same data gives different values.

Since they differ, you think that the values are computed before the training step ? I will save the model at step 29 to check this out.

**Comment by eduardofv** [comment]
Those values should be calculated _after_ the forward but _before_ the back, that is, before updating weights based on the gradients.If you save at step 29, the same would apply: you will have the loss at that set but the model will be updated anyway. I've just come to this [tweet](https://twitter.com/fchollet/status/1132013928681050113) by @fchollet in which he shows how to implement a custom version of the fit step: calculate predictions, calculate loss, calculate gradients, update weights. Maybe you could calculate again the loss after the update and show. My guess is that you should get same results as in evaluate.

**Comment by eduardofv** [comment]
I've done something like that [here](https://colab.research.google.com/drive/1rBtQ-IlhEILU4fYpj8zM1XcayxR8x9CW). From the results you can see that the initial loss is the same as the "post-loss" at the previous step.

```
loss_fn = tf.keras.losses.CategoricalCrossentropy()
optimizer = tf.keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999)

@tf.function
def custom_fit():
  with tf.GradientTape() as tape:
    predictions = model(x_train.astype("float32"))
    loss_value = loss_fn(y_train, predictions)
   
  grads = tape.gradient(loss_value, model.trainable_weights)
  optimizer.apply_gradients(zip(grads, model.trainable_weights))
  
  post_predictions = model(x_train.astype("float32"))
  post_loss = loss_fn(y_train, post_predictions)
  return (loss_value, post_loss)

for i in range(10):
  lv = custom_fit()
  print(lv)

(<tf.Tensor: id=914, shape=(), dtype=float32, numpy=52.713608>, <tf.Tensor: id=915, shape=(), dtype=float32, numpy=72.112686>)
(<tf.Tensor: id=918, shape=(), dtype=float32, numpy=72.112686>, <tf.Tensor: id=919, shape=(), dtype=float32, numpy=73.35305>)
(<tf.Tensor: id=922, shape=(), dtype=float32, numpy=73.35305>, <tf.Tensor: id=923, shape=(), dtype=float32, numpy=55.93804>)
(<tf.Tensor: id=926, shape=(), dtype=float32, numpy=55.93804>, <tf.Tensor: id=927, shape=(), dtype=float32, numpy=38.91731>)
(<tf.Tensor: id=930, shape=(), dtype=float32, numpy=38.91731>, <tf.Tensor: id=931, shape=(), dtype=float32, numpy=27.62747>)
(<tf.Tensor: id=934, shape=(), dtype=float32, numpy=27.62747>, <tf.Tensor: id=935, shape=(), dtype=float32, numpy=19.682486>)
(<tf.Tensor: id=938, shape=(), dtype=float32, numpy=19.682486>, <tf.Tensor: id=939, shape=(), dtype=float32, numpy=13.982845>)
(<tf.Tensor: id=942, shape=(), dtype=float32, numpy=13.982845>, <tf.Tensor: id=943, shape=(), dtype=float32, numpy=10.246971>)
(<tf.Tensor: id=946, shape=(), dtype=float32, numpy=10.246971>, <tf.Tensor: id=947, shape=(), dtype=float32, numpy=8.010065>)
(<tf.Tensor: id=950, shape=(), dtype=float32, numpy=8.010065>, <tf.Tensor: id=951, shape=(), dtype=float32, numpy=8.71225>)
```

**Comment by julj** [comment]
You're right, the metrics monitored are computed before the weight update.
I wrote a custom monitoring callback to find out:

```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import mnist
from tensorflow.keras.callbacks import Callback


config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)


#################################
# Generate fake multilabel data #
#################################

batch_size = 1024
num_classes = 10
h, w = 28, 28

(x_train, y_train), (_, _) = mnist.load_data()
x_train = x_train[:1024]
x_train = x_train.reshape(x_train.shape[0], h, w, 1) # default tf format is NHWC
y_train = y_train[:1024]
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
    

###############
# build model #
###############

inputs = keras.Input(shape=(h, w, 1), batch_size=batch_size)
x = layers.Conv2D(32, (3, 3))(inputs)
x = layers.Activation('relu')(x)
x = layers.MaxPooling2D(pool_size=(3, 3), strides=(2,2))(x)
x = layers.Flatten()(x)
x = layers.Dense(100, activation='elu', name='logits')(x)
outputs = layers.Dense(num_classes, activation='softmax')(x)
    
model = tf.keras.Model(inputs=inputs, outputs=outputs)

model.summary()


################################################
# define training parameters and compile model #
################################################

model.compile(
    loss=keras.losses.categorical_crossentropy,
    optimizer=keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999),
    metrics=['accuracy']
)


#####################
# custom monitoring #
#####################


class CustomMonitoring(Callback):

    def on_train_batch_end(self, batch, logs=None):
        loss, acc = self.model.evaluate(x_train, y_train, batch_size=1024)
        print('For end batch {}, loss is {:7.2f} and acc is {}.'.format(batch, loss, acc))


#########
# train #
#########

print("***** Training *****")

model.fit(
    x_train,
    y_train,
    epochs=10,
    verbose=2,
    callbacks=[CustomMonitoring()]
)


############
# Evaluate #
############

print("***** Evaluation *****")

ev = model.evaluate(x_train, y_train, batch_size=1024)
print(ev)
```
Output:

```
Epoch 1/10
1024/1024 [==============================] - 0s 24us/sample - loss: 120.1128 - acc: 0.1514
For end batch 0, loss is  120.11 and acc is 0.1513671875.
1024/1024 - 1s - loss: 59.1023 - acc: 0.1035
Epoch 2/10
1024/1024 [==============================] - 0s 4us/sample - loss: 101.6853 - acc: 0.2588
For end batch 0, loss is  101.69 and acc is 0.2587890625.
1024/1024 - 0s - loss: 120.1128 - acc: 0.1514
Epoch 3/10
1024/1024 [==============================] - 0s 4us/sample - loss: 108.9812 - acc: 0.2500
For end batch 0, loss is  108.98 and acc is 0.25.
1024/1024 - 0s - loss: 101.6853 - acc: 0.2588
Epoch 4/10
1024/1024 [==============================] - 0s 4us/sample - loss: 93.5093 - acc: 0.3691
For end batch 0, loss is   93.51 and acc is 0.369140625.
1024/1024 - 0s - loss: 108.9812 - acc: 0.2500
Epoch 5/10
1024/1024 [==============================] - 0s 4us/sample - loss: 85.9527 - acc: 0.2939
For end batch 0, loss is   85.95 and acc is 0.2939453125.
1024/1024 - 0s - loss: 93.5093 - acc: 0.3691
Epoch 6/10
1024/1024 [==============================] - 0s 4us/sample - loss: 65.5312 - acc: 0.4141
For end batch 0, loss is   65.53 and acc is 0.4140625.
1024/1024 - 0s - loss: 85.9527 - acc: 0.2939
Epoch 7/10
1024/1024 [==============================] - 0s 3us/sample - loss: 59.6325 - acc: 0.3525
For end batch 0, loss is   59.63 and acc is 0.3525390625.
1024/1024 - 0s - loss: 65.5312 - acc: 0.4141
Epoch 8/10
1024/1024 [==============================] - 0s 4us/sample - loss: 52.5296 - acc: 0.3838
For end batch 0, loss is   52.53 and acc is 0.3837890625.
1024/1024 - 0s - loss: 59.6325 - acc: 0.3525
Epoch 9/10
1024/1024 [==============================] - 0s 4us/sample - loss: 48.6921 - acc: 0.3516
For end batch 0, loss is   48.69 and acc is 0.3515625.
1024/1024 - 0s - loss: 52.5296 - acc: 0.3838
Epoch 10/10
1024/1024 [==============================] - 0s 4us/sample - loss: 39.5250 - acc: 0.3945
For end batch 0, loss is   39.53 and acc is 0.39453125.
1024/1024 - 0s - loss: 48.6921 - acc: 0.3516
***** Evaluation *****
1024/1024 [==============================] - 0s 4us/sample - loss: 39.5250 - acc: 0.3945
[39.5250129699707, 0.39453125]

```

**Comment by eduardofv** [comment]
Great!

**Comment by XXZhou25** [comment]
This is super helpful. Thank you guys! I also had this confusion but solved now.

**Comment by rubenlg** [comment]
Thanks for filing this issue. I just had this problem as well. Reading the reason why it happens, it makes a lot of sense. In order to report the metrics after the backpropagation, tensorflow would need to do an extra forward pass, which has performance implications.
